### PR TITLE
Poll Twitter for tweets made containing a given hashtag

### PR DESCRIPTION
**Pull request by michaeldyrynda**
Addresses #5.

For any found tweets, add them as a new unflagged topic against the owner user record, so that they can be reviewed.

I've done some preliminary work on this and opened a PR so I can get some feedback on the direction I've taken so far. The basic theory is that a user will specify a hashtag that Suggestive looks for periodicially (scheduling the `suggestive:poll-twitter` command), and creates unflagged `Topic` records for the owner to review.

I assume as well that the `/topics` API endpoint should return only accepted topics (it currently returns everything).

I haven't yet made any interface to set a hashtag, though that's only because there's no user profile at this stage.

This will also use the application Twitter credentials, which is 150 requests every ~15 minutes. This may be ok initially, but it is trivial to make it 150 requests per user - you just need to capture the Twitter token and secret when the user signs up. Of course, that's not much good with a GitHub authenticated user, so you might want to lock it down to just Twitter users.

### CODE REVIEW

**Comment by mattstauffer** [comment]
Quick note: 

> I assume as well that the /topics API endpoint should return only accepted topics (it currently returns everything).

Nope. https://github.com/mattstauffer/suggestive/blob/master/tests/AdminTopicTest.php#L74 You can pass in parameters; haven't finalized what they'll be but right now it would be `api/topics?status=accepted`

**Comment by michaeldyrynda** [comment]
I saw that part. I meant to say that the topics list currently shows topics whether they're flagged as accepted or not. 

The [dashboard](https://github.com/mattstauffer/suggestive/blob/master/resources/assets/js/app.js) currently hits `/topics` without any query parameters.

**Comment by mattstauffer** [comment]
@deringer Aha, got it :) :+1:

will review PR later, busy day w the fam today :)

File: resources/assets/js/app.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ var App = Vue.extend({
         };
     },
     created: function () {
-        this.$http.get('topics', function (data, status, request) {
+        this.$http.get('topics?status=accepted', function (data, status, request) {
```
Comment (mattstauffer) [inline resources/assets/js/app.js]: Nope, we actually want to list out all proposed topics for the purpose of voting.

File: app/Console/Commands/PollTwitter.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace App\Console\Commands;
+
+use App\User;
+use Illuminate\Console\Command;
+use Log;
```
Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php]: When using a Façade, it's preferred to pull in the full Façade route

File: app/Console/Commands/PollTwitter.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+     * @return mixed
+     */
+    public function handle()
+    {
+        $users = User::where('role', 'owner')
+                     ->whereNotNull('hashtag')
+                     ->get([ 'id', 'name', 'hashtag', 'last_tweet_id', ]);
+
+        if ($users) {
+            $twitter = app('TwitterClient');
```
Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php]: You can inject this into the command, afaik

File: app/Console/Commands/PollTwitter.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+            Log::info('Found owners to poll Twitter for', [ 'count' => $users->count(), ]);
+
+            $users->each(function ($user) use ($twitter) {
+                Log::info('Polling Twitter', [ 'owner' => $user->name, 'hashtag' => $user->hashtag, ]);
```
Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php]: No need for spaces inside of array

Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php]: Also, the trailing comma is unnecessary on single-line arrays

File: app/Console/Commands/PollTwitter.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+                $response = $twitter->get('search/tweets', [ 'q' => $user->hashtag, 'since_id' => $user->last_tweet_id, ]);
```
Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php]: Is there no `since_datetime` instead of `since_id`? would speed up the check.. if no matching tweets for two years, we have to check two years this way

File: app/Console/Commands/PollTwitter.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        $users = User::where('role', 'owner')
+                     ->whereNotNull('hashtag')
+                     ->get([ 'id', 'name', 'hashtag', 'last_tweet_id', ]);
+
+        if ($users) {
+            $twitter = app('TwitterClient');
+
+            Log::info('Found owners to poll Twitter for', [ 'count' => $users->count(), ]);
+
+            $users->each(function ($user) use ($twitter) {
```
Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php]: I would likely turn this into a queuable `Job`

File: nitpick.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
+  "ignore": [
+    "tests/*",
```
Comment (mattstauffer) [inline nitpick.json]: Actually, I would rather we don't ignore these folders. I know it means we get namespace complaints everywhere, but I still like getting the PSR-2 sniff even if it means haaving to delete all the namespace complaints manually.

File: package.json [hunk 8]
```diff
@@ -1,11 +1,13 @@
 {
   "private": true,
   "devDependencies": {
+    "babel-preset-react": "^6.3.13",
```
Comment (mattstauffer) [inline package.json]: Why? Isn't this a dependency of something?

**Comment by mattstauffer** [comment]
Looks good. Architecturally, I wonder if it wouldn't be better to have something of a "machine" user that could be the one that gets assigned these topics?

File: resources/assets/js/app.js (hunk 1 above)
Comment (michaeldyrynda) [inline resources/assets/js/app.js]: Ahhh, my bad - I misunderstood the purpose of the accepted status.

File: package.json (hunk 8 above)
Comment (michaeldyrynda) [inline package.json]: Right you are, I didn't check - just went based on the feedback in #11. I had some issues compiling the `app.js`, but it turns out that was probably just NPM-related.

File: nitpick.json (hunk 7 above)
Comment (michaeldyrynda) [inline nitpick.json]: Roger

File: app/Console/Commands/PollTwitter.php (hunk 5 above)
Comment (michaeldyrynda) [inline app/Console/Commands/PollTwitter.php]: Sadly [not](https://dev.twitter.com/rest/reference/get/search/tweets). Twitter lets you search for tweets `until` a given date, or since a given id, handling the filtering on their end, based on the query. This is why at the end of the process we store the first (most recent) returned response to search after.

Also, the API will return only `<limit>` (currently 100) results at a time, so what you specify as the `since_id` will be forced to the oldest available.

File: app/Console/Commands/PollTwitter.php (hunk 3 above)
Comment (michaeldyrynda) [inline app/Console/Commands/PollTwitter.php]: It's not an interface, it's an (I don't know what the proper name is) alias to the underlying return, given the `TwitterOAuth` client doesn't implement an interface.

Would you suggest that in this instance we would create an app-specific wrapper for it that implements an interface so that we can bind to it?

This particular solution has always seemed hacky to me (I use the same method in another app I have). The alternative would be to wrap it in a facade?

File: app/Console/Commands/PollTwitter.php (hunk 2 above)
Comment (michaeldyrynda) [inline app/Console/Commands/PollTwitter.php]: I didn't know that was possible - did Laravel always used to work like that?

File: app/Console/Commands/PollTwitter.php (hunk 6 above)
Comment (michaeldyrynda) [inline app/Console/Commands/PollTwitter.php]: What's the theory behind `Job` vs scheduled `Command`? What would the trigger be to the queue the job to execute? Or are you just suggesting that I queue the request and response part of the command so that it doesn't need to handle that part of it?

**Comment by michaeldyrynda** [comment]
Some kind of system user was my first thought, but I don't know what the multi-tenant vision is for Suggestive, so wasn't sure how a polled status would then be assigned back to the appropriate owner. Happy to discuss and implement!

File: app/Console/Commands/PollTwitter.php [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    public function handle()
+    {
+        $users = User::where('role', 'owner')
+                     ->whereNotNull('hashtag')
+                     ->get(['id', 'name', 'hashtag', 'last_tweet_id']);
+
+        if ($users) {
+            Log::info('Found owners to poll Twitter for', ['count' => $users->count()]);
+
+            $twitter = app('TwitterClient');
```
Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php]: Can inject?

File: public/js/app.js [hunk 10]
```diff
@@ -13295,7 +13390,7 @@ if (module.hot) {(function () {  module.hot.accept()
   var hotAPI = require("vue-hot-reload-api")
   hotAPI.install(require("vue"), true)
   if (!hotAPI.compatible) return
-  var id = "/Users/mattstauffer/Sites/suggestive/resources/assets/js/components/user-dashboard.vue"
+  var id = "/Users/michael/Development/suggestive/resources/assets/js/components/user-dashboard.vue"
```
Comment (mattstauffer) [inline public/js/app.js:13393]: lol

File: app/Console/Commands/PollTwitter.php (hunk 9 above)
Comment (michaeldyrynda) [inline app/Console/Commands/PollTwitter.php]: Oh man, I commented on your comments above but GitHub has hidden them all now 😁

There's a couple of others where I asked for feedback on, too. Should have waited before pushing new commits. I'm not a GitHub noob >_>

File: public/js/app.js (hunk 10 above)
Comment (michaeldyrynda) [inline public/js/app.js:13393]: Ohhhh boy, that's not ideal

File: app/Console/Commands/PollTwitter.php (hunk 9 above)
Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php]: No, no need to wait; usually it won't wipe the old comments. I'll look through them.

On this one, your comment was:

> It's not an interface, it's an (I don't know what the proper name is) alias to the underlying return, given the TwitterOAuth client doesn't implement an interface.
> 
> Would you suggest that in this instance we would create an app-specific wrapper for it that implements an interface so that we can bind to it?
> 
> This particular solution has always seemed hacky to me (I use the same method in another app I have). The alternative would be to wrap it in a facade?

I'm not sure what you mean for the Interface; why don't you just typehint `Abraham\TwitterOAuth\TwitterOAuth`? If you want `TwitterClient` as a shortcut like you have it here, you can bind one of them like you are right now, and then have the other one alias to it. Ping me here if you want me to write out the code for what that would look like.

File: app/Console/Commands/PollTwitter.php [hunk 11]
```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace App\Console\Commands;
+
+use App\User;
+use Illuminate\Console\Command;
+use Illuminate\Support\Facades\Log;
```
Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php:10]: Your question originally here was:

> I didn't know that was possible - did Laravel always used to work like that?

I never actually need to worry about it until 5.0, so I can only speak to it working that way back to 5.0, but I dont' know of any significant changes to the structure of Façades for quite a while, so I would assume yes?

File: app/Console/Commands/PollTwitter.php [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+                     ->whereNotNull('hashtag')
+                     ->get(['id', 'name', 'hashtag', 'last_tweet_id']);
+
+        if ($users) {
+            Log::info('Found owners to poll Twitter for', ['count' => $users->count()]);
+
+            $twitter = app('TwitterClient');
+
+            $users->each(function ($user) use ($twitter) {
+                Log::info('Polling Twitter', ['owner' => $user->name, 'hashtag' => $user->hashtag]);
```
Comment (mattstauffer) [inline app/Console/Commands/PollTwitter.php]: I suggested queue a job here. Your question was:

> What's the theory behind Job vs scheduled Command? What would the trigger be to the queue the job to execute? Or are you just suggesting that I queue the request and response part of the command so that it doesn't need to handle that part of it?

I would recommend something like this:

``` php
$users->each(function ($user) {
    $this->dispatch(new PollTwitterForUsersHashtag($user));
}
```

The benefit of separating the logic here is that you A) have encapsulated a single user's poll logic in the same way this artisan command is encapsulating a single cron-ready action of "do all the polls for this site", and B) if the twitter pulls are slow or hitting rate limits or whatever, each individual user's pull is separate and queued, which means they're not holding up the whole stack... you can throw multiple queue listeners.. you can delete them after a certain number of failures... etc. 

Basically, queuing tasks like this gives you a much greater level of control and flexiblity.

In response to your question, I'm still 100% in favor of making this an artisan command, but I just like it to be a command that then queues jobs.

File: app/Console/Commands/PollTwitter.php (hunk 9 above)
Comment (michaeldyrynda) [inline app/Console/Commands/PollTwitter.php]: Yeah, for some reason I had trouble getting this to work in another project use used `TwitterOAuth` with, but that's because I was passing the additional parameters to it, so needed to use `app()->make()` to pass the extra parameters in.

I'll push an update shortly that fixes this and pushes the functionality off to a `Job` :+1:

File: app/Console/Commands/PollTwitter.php (hunk 5 above)
Comment (michaeldyrynda) [inline app/Console/Commands/PollTwitter.php]: Turns out the application limit is 450 requests per `<period>`, which is currently 15 minutes.

It's 150 requests per user per `<period>`, if we were to make requests for each user i.e. with their token and secret.

That said, 15 minutes should be an ok interval to go buy, and until you're looking at hundreds of users, I wouldn't think you'd hit the per-application limit, anyway.

**Comment by mattstauffer** [comment]
@deringer I OWE YOU A PR REVIEW OF THIS. I still remember!

**Comment by michaeldyrynda** [comment]
In your own time my friend, though I’ll need to leverage some of the code I wrote here on another project at some point so another set of eyes over it will be handy (no rush!)

Michael Dyrynda

Sent from
https://polymail.io/

On Wed, 02 Mar 2016 at 3:12 PM Matt Stauffer

<
mailto:Matt Stauffer notifications@github.com

> wrote:

https://github.com/deringer
I OWE YOU A PR REVIEW OF THIS. I still remember!

**Comment by mattstauffer** [comment]
@michaeldyrynda holy crap i never reviewed this :/ // / / / /// // // // /

**Comment by michaeldyrynda** [comment]
I did look at this yesterday. Now there are lots merge conflicts with compiled assets lol
